Storage/REST-API.md

```diff
@@ -0,0 +1,183 @@
+
+* **PUT**: Create a new record for a resource, or update all fields of an existing resource.
+
+  Attempting to `PUT` to a URL that has an incomplete view of a record would result in a `405 Method Not Allowed` error.
+  
+  In the `&format=full` view, `PUT` is equivalent to `PATCH`. 
+  
+  `PUT` behaves the same way as `PATCH` when it comes to avoiding merge conflicts.
+
+  When legal, the only difference between `PUT` and `PATCH` is that `PUT` will create a new resource if it did not exist before the call. A `201 Created` response code will then be returned. Use `PATCH` if this behavior is undesirable.
+
+  `TODO: Location header https://tools.ietf.org/html/rfc7231#section-4.3.3`
+
+* **POST**: Create a new record and assign it to some recourse.
+
+  `POST` will create a new resource. The URL for POST must not contain the key(s) for the record being posted, and the payload body should not contain the key(s) either.
+
+  The generation of the new key is user-defined. Three common strategies are: 1) random, 2) the [salted] hash of the payload, 3) the value of certain passed in field or fields.
+  
+  In the latter two, as well as other, cases, a `POST` request may result in an attempt to overwrite an already existing resource. By default, `POST` will not overwrite, but return a `409 Conflict` error code, with the top-level `"url"` field of the returned JSON object containing the URL for the resource that was about to be overwritten.
+
+  The `&overwrite=1` / `&overwrite=true` URL parameter can be set to indicate that the `POST` request should complete successfully even if it would result in overwriting existing data.
+
+Successful `POST` results in either `200 (OK)` error code if the resource has been updated, or `201 (Created)`.
+
+* **DELETE**: Deletes the resource.
+
+   Similarly to `PATCH` and `PUT`, `DELETE` should be called on the URL that has been (or could have been) returned as the top-level `"url":"..."` field of the payload of `GET`.
+   
+   `DELETE` follows the same conflict resolution rules as `PATCH` and `PUT`.
+   
+`TODO: User-defined data integrity checks, and what errors are returned if they fail?`
+
+
+### Discoverability
+
+The entry point (usually, `"/"`) URL will contain the list of inner `"url_*"`-s to access respective fields ("tables") of the storage.
+
+The entry point URL may also contain the URLs by which the user can access the respective schema files. This is applicable to F# and C++.
+
+[ TO DISCUSS ] The payload when `GET`-ting the resource corresponding to an individual record (ex. `/user/123`) will contain the `{...,"url_collection":".../users"}` reference to the URL to browse the whole collection, the record from which is being requested.
```

Comment:
Out of context, I'm leaning towards "no, because de-denormalization". The resource belongs to one collection; other resources may point to it.

In the RESTful world, however, when there is a reference from one resource to another, instead of just returning the URL, we may return the object containing the `preview` format (which obviously includes the `.url` field too).